core/src/raw/oio/read/buffer_reader.rs

```diff
@@ -31,6 +31,76 @@ use crate::*;
 
 use super::BlockingRead;
 
+/// [MaybeBufferReader] is a wrapper to fill the gap between reader and buffer reader.
+pub enum MaybeBufferReader<R> {
+    /// Use the Reader directly.
+    NoBuffer(R),
+    /// Wrap with the [BufferReader].
+    Buffer(BufferReader<R>),
+}
+
+impl<R> MaybeBufferReader<R> {
+    /// Returns a factory to construct a [MaybeBufferReader].
+    pub fn factory(cap: Option<usize>) -> impl FnOnce(R) -> MaybeBufferReader<R> {
+        move |r| match cap {
+            Some(size) => MaybeBufferReader::Buffer(BufferReader::new(r, size)),
+            None => MaybeBufferReader::NoBuffer(r),
+        }
+    }
+}
```

Comment:
Finally, I realized we might not need this high-order function. Maybe just use the `MaybeBufferReader::new(r,cap)`.